src/core/server/http/http_service.ts

```diff
@@ -150,4 +169,28 @@ export class HttpService implements CoreService<HttpServiceSetup, HttpServiceSta
     await Promise.all([...this.secondaryServers.values()].map(s => s.stop()));
     this.secondaryServers.clear();
   }
+
+  private async runNotReadyServer(config: HttpConfig) {
+    const httpServer = new HttpServer(this.log);
+    const { server } = await httpServer.setup(config);
+    this.notReadyServer = server;
+    // use hapi server while Kibana ResponseFactory doesn't allow specifying custom headers
+    // https://github.com/elastic/kibana/issues/33779
+    this.notReadyServer.route({
+      path: '/{p*}',
+      method: '*',
+      handler: (req, responseToolkit) => {
+        this.log.debug(`Kibana server is not ready yet ${req.method}:${req.url}.`);
+
+        // If server is not ready yet, because plugins or core can perform
+        // long running tasks (build assets, saved objects migrations etc.)
+        // we should let client know that and ask to retry after 30 seconds.
+        return responseToolkit
+          .response('Kibana server is not ready yet')
+          .code(503)
+          .header('Retry-After', '30');
+      },
+    });
+    await this.notReadyServer.start();
```

Comment:
There is no catch here, so you can return this promise directly without the `await`:

```js
return this.notReadyServer.start();
```

Reply:
I don't want to return a start contract

Reply:
Not related to this particular case, but I remember we had a long debate in platform/ops teams on  `async () => return something-that-returns-promise` vs `async () => return await something-that-returns-promise`. Did we settle on something? :) Would be great if we all agree on just one approach (too bad we don't have a linter for that).

Reply:
there is no linter rule for this, but we kinda follow the convention `async () => await promise` to highlight the fact it is a promise

Reply:
I prefer `return await promiseThing` to make it explicit that if you want to try/catch what this thing does, you need to await it. I'll open an issue for adding a lint rule for this.

Reply:
👍 for a linter rule,  https://eslint.org/docs/rules/require-await